configure.ac

```diff
@@ -1768,14 +1768,23 @@ fi
 # Check for --enable-experimental-jit:
 AC_MSG_CHECKING([for --enable-experimental-jit])
 AC_ARG_ENABLE([experimental-jit],
-              [AS_HELP_STRING([--enable-experimental-jit],
+              [AS_HELP_STRING([--enable-experimental-jit@<:@=no|yes|yes-default-off|interpreter@:>@],
                               [build the experimental just-in-time compiler (default is no)])],
               [],
               [enable_experimental_jit=no])
+case $enable_experimental_jit in
+  no) enable_experimental_jit=no ;;
+  yes) enable_experimental_jit="-D_Py_JIT -D_Py_TIER2=1" ;;
+  yes-default-off) enable_experimental_jit="-D_Py_JIT -D_Py_TIER2=3" ;;
+  interpreter) enable_experimental_jit="-D_Py_TIER2=4" ;;
+  interpreter-default-off) enable_experimental_jit="-D_Py_TIER2=6" ;;  # Secret option
+  *) AC_MSG_ERROR(
+      [invalid argument: --enable-experimental-jit=$enable_experimental_jit; expected no|yes|yes-default-off|interpreter]) ;;
```

Comment:
I find it a bit hard to reason about the relationship between `_Py_JIT` and the `_Py_TIER2` flags... I think there's redundant information here.

Maybe it would be clearer to just use `_Py_JIT` and have the different options be encoded in its bits. *Or* have orthogonal defines and get rid of the bit masks entirely:

- `_Py_JIT`: Defined for `yes`, `yes-off`.
- `_Py_JIT_ENABLED`: Defined for `yes` and `interpreter`.
- `_Py_JIT_INTERPRETER`: Defined for `interpreter` and `interpreter-off`.

(I *think* these make sense, in terms of usefulness?)

Again, this stuff is all internal, so we can tweak it if need be.

Reply:
I tried that and something broke (a segfault in `_build_python`) and I couldn't figure out what I'd done wrong. All the JIT stuff that you're responsible for keeps using `#ifdef _Py_JIT`. All the stuff that is shared between the JIT and the Tier 2 interpreter checks for `_Py_TIER2` -- almost all in the form of `#ifdef _Py_TIER2`. There are only 2 places that have a different form of test, one in ceval.c (L758, where some variables only exist for the interpreter) and one in pylifecycle.c (checking `_Py_TIER2 & 2` to determine if the default should be on or off).

Reply:
Actually, what I tried was folding the bitmask into a single `_Py_JIT` flag, and that definitely was no improvement. I am trying to limit the number of predefined variables; apparently 1 didn't work, and I like how it turned out with 2 (JIT and TIER2), so I'm sticking with what I've got.

As you say, this is all internal and nobody else should care, so it should be easy to change later.